### PR TITLE
Check task call inputs

### PR DESCRIPTION
**Pull request by aednichols**
Partially addresses https://github.com/broadinstitute/winstanley/issues/83

There are a few limitations caused by other issues; I chose to stop here and ship instead of solving them upfront. They are noted in the `ERRATA` section of the test file (`call_input_checking.wdl)`.

### CODE REVIEW

File: tests/testData/annotation/call_input_checking.wdl [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    # No inputs, none provided (correct)
+    call no_inputs
+
+    # No inputs, empty scope for input block (highlight)
+    <error descr="Task does not take inputs.">call no_inputs {
+
+    }</error>
+
+    # No inputs, empty input list (highlight)
+    # IMO: this one is a bit debatable as to whether it should highlight, since the list IS technically empty (AEN, 2018-07-17)
```
Comment (aednichols) [inline tests/testData/annotation/call_input_checking.wdl]: This depends on some combination of what the spec says and what Cromwell does (need to check)

File: src/winstanley/WdlAnnotator.scala [hunk 2]
```diff
@@ -46,10 +48,10 @@ class WdlAnnotator extends Annotator {
     case value: WdlCallableLookup =>
       // TODO (Issue 63): Work out a better way to verify that imports exist and (if local?) that an appropriate task/workflow inside them exist.
       // For now to avoid error highlights, I'll just suppress errors for things that look like FQNs
-      if (value.getFullyQualifiedName.getNode.getChildren(null).length == 1) {
+      if (value.getFullyQualifiedName.getNode.getChildren(TokenSet.ANY).length == 1) {
```
Comment (aednichols) [inline src/winstanley/WdlAnnotator.scala:51]: I'm pretty sure I've correctly interpreted the meaning of `getChildren(null)` to be "get children of any type", but could use a double-check (doesn't seem to be documented by JetBrains)

File: tests/testData/annotation/call_input_checking.wdl [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+    }</error>
+
+    # Empty input list (highlight)
+    <error descr="Missing required inputs(s) for task: a, b, c">call required_inputs {
+        input:
+    }</error>
+
+    # Missing inputs (highlight)
+    <error descr="Missing required inputs(s) for task: b, c">call required_inputs {
```
Comment (cjllanwarne) [inline tests/testData/annotation/call_input_checking.wdl]: A "missing input" is not an error and should be highlighted as a warning, with text something like:
> Incomplete workflow interface: inputs 'File a' and 'String b' to call 'foo' are not provided by the workflow. To use as a top-level workflow inputs 'wf.foo.a' and 'wf.foo.b' will be required. To use this workflow as a subworkflow, you'll need to pass through workflow inputs for these values.

Side note: this message would be the place to add the auto-completion option in #94.

Comment (aednichols) [reply tests/testData/annotation/call_input_checking.wdl]: Is it wise to show this warning if we don't give the user any way to make it go away?

We can't exactly check the user's input JSON.

File: src/winstanley/WdlAnnotator.scala (hunk 2 above)
Comment (cjllanwarne) [reply src/winstanley/WdlAnnotator.scala:51]: I'm happy to let the tests passing be the judge of "is this valid", and it seems so.

File: src/winstanley/WdlAnnotator.scala [hunk 4]
```diff
@@ -69,10 +71,81 @@ class WdlAnnotator extends Annotator {
...
+      }
     case _ => ()
   }
 
   private def kvName(kv: WdlKv): Option[String] = Option(kv.getNode.findChildByType(WdlTypes.IDENTIFIER)).map(_.getText)
   private def runtimeKeyword(r: WdlRuntimeBlock): Option[ASTNode] = Option(r.getNode.findChildByType(WdlTypes.RUNTIME))
   private def outputKeyword(r: WdlTaskOutputs): Option[ASTNode] = Option(r.getNode.findChildByType(WdlTypes.OUTPUT))
+
+  private def compareInputListsForCall(psiElement: PsiElement, annotationHolder: AnnotationHolder, callInput: WdlCallInput, taskInputSection: WdlTaskSection) = {
+    // In the interest of unambiguous naming, this method (ab)uses the formal definitions of parameter (declared on method) and argument (passed to method)
```
Comment (cjllanwarne) [inline src/winstanley/WdlAnnotator.scala:116]: I like this distinction... to make this just as clear in the variable names could they be, for example `taskParameters` and `callArguments`?

File: src/winstanley/WdlAnnotator.scala [hunk 5]
```diff
@@ -69,10 +71,81 @@ class WdlAnnotator extends Annotator {
...
+        val taskName = callBlock.getCallableLookup.getName
+        val tasksInScope: Set[WdlTaskBlockImpl] = callBlock.findTasksInScope
+
+        tasksInScope.find(_.getTaskDeclaration.getName == taskName)
+      }
+
+      taskForCall(callBlock) match {
+        case Some(task) =>
+          // The "type" of a WdlTaskSection is determined by which one of its getXXX() methods returns not-null. Checkmate, Scala nerds.
+          val maybeTaskInputSection: Option[WdlTaskSection] = task.getTaskSectionList.asScala.find(_.getInputBlock != null)
```
Comment (cjllanwarne) [inline src/winstanley/WdlAnnotator.scala]: Does this also work in draft-2 WDLs which don't use input blocks?

File: tests/testData/annotation/call_input_checking.wdl [hunk 6]
```diff
@@ -0,0 +1,236 @@
+version 1.0
```
Comment (cjllanwarne) [inline tests/testData/annotation/call_input_checking.wdl]: If we do this check in draft-2, we should have a draft-2 version of this test file.

**Review by cjllanwarne — APPROVED** [review]
Approve, pending double-check that it works as expected in draft-2 too.

File: tests/testData/annotation/no_call_input_checking_draft_2.wdl [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    # Duplicate input (correct; would be an error highlight in 1.0)
+    call required_inputs {
+        input:
+            a = "a",
+            a = "a",
+            b = 5,
+            c = false
+    }
+
+    # Unsupplied inputs (correct; woudl be a warning highlight in 1.0)
```
Comment (cjllanwarne) [inline tests/testData/annotation/no_call_input_checking_draft_2.wdl]: Typo on `woudl`

File: tests/testData/annotation/no_call_input_checking_draft_2.wdl [hunk 8]
```diff
@@ -0,0 +1,35 @@
+# Exists to make sure we do NOT highlight bad inputs in draft-2
+
+workflow my_workflow {
+
+    # Duplicate input (correct; would be an error highlight in 1.0)
```
Comment (cjllanwarne) [inline tests/testData/annotation/no_call_input_checking_draft_2.wdl]: Rather than `correct`, maybe just `no annotation` is more accurate?

**Review by cjllanwarne — APPROVED** [review]
👍 assuming you've run the full set of tests and they all went green

**Pull request merged**